bson/encoding.go

```diff
@@ -0,0 +1,231 @@
+package bson
+
+//  all the methods relative to the set/get json/bson should be stored
+//  in this file.
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/keep-network/paillier"
+	"gopkg.in/mgo.v2/bson"
+)
+
+type Cypher paillier.Cypher
+type ThresholdPublicKey paillier.ThresholdPublicKey
+type PublicKey paillier.PublicKey
+type PrivateKey paillier.PrivateKey
+type PartialDecryptionZKP paillier.PartialDecryptionZKP
+
+type dbCypher struct {
+	C string
+}
+
+func (cypher *Cypher) GetBSON() (interface{}, error) {
+	return &dbCypher{fmt.Sprintf("%x", cypher.C)}, nil
+}
+
+func (cypher *Cypher) SetBSON(raw bson.Raw) error {
+	c := dbCypher{}
+	if err := raw.Unmarshal(&c); err != nil {
+		return err
+	}
+	var ok bool
+	cypher.C, ok = new(big.Int).SetString(c.C, 16)
+	if !ok {
+		return errors.New("big int not in hexadecimal format")
+	}
+	return nil
+}
```

Comment:
In order to avoid conversion between `Cypher` and `paillier.Cypher` in code using this API, I'd propose to do the following:
```
func SerializeCypher(cypher *paillier.Cypher) ([]byte, error) {
	return bson.Marshal(toSerializableCypher(cypher))
}

func DeserializeCypher(data []byte) (*paillier.Cypher, error) {
	serializable := new(Cypher)
	if err := bson.Unmarshal(data, serializable); err != nil {
		return nil, err
	}

	return toOriginalCypher(serializable), nil
}

func toSerializableCypher(cypher *paillier.Cypher) *Cypher {
	serializable := Cypher(*cypher)
	return &serializable
}

func toOriginalCypher(serializable *Cypher) *paillier.Cypher {
	original := paillier.Cypher(*serializable)
	return &original
}
```

This way we define a clear API - `SerializeCypher` and `DeserializeCypher` as well as hide implementation details behind it.

Then, the test is just:
```
func TestCypherBsonSerialization(t *testing.T) {
	cypher := &paillier.Cypher{
		C: b(5),
	}

	serialized, err := SerializeCypher(cypher)
	if err != nil {
		t.Fatal(err)
	}

	deserialized, err := DeserializeCypher(serialized)
	if err != nil {
		t.Fatal(err)
	}

	if !reflect.DeepEqual(cypher, deserialized) {
		t.Errorf(
			"Unexpected serialization result\nActual: %v\nExpected: %v\n",
			deserialized,
			cypher,
		)
	}
}
```

Reply:
^ same for all other types we let to serialize

Reply:
Refactored it this way you mentioned. Added methods Serialize* Deserialize* for all types. Also extracted them from `encoding.go` to separate files per type.